### PR TITLE
auto_memory_nodeset_placement: add case auto mem nodeset and placement

### PR DESCRIPTION
**Pull request by dzhengfy**
Depend on
- https://github.com/avocado-framework/avocado-vt/pull/3716
- https://github.com/avocado-framework/avocado-vt/pull/3722
- https://github.com/autotest/tp-libvirt/pull/5034

Case ID: xxx-296727

Signed-off-by: Dan Zheng <dzheng@redhat.com>

### CODE REVIEW

**Comment by dzhengfy** [comment]
libvirt 9.2.0

`
 (01/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_defined.mem_mode_strict.vcpu_auto: PASS (38.72 s)
 (02/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_defined.mem_mode_strict.vcpu_static: PASS (41.63 s)
 (03/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_defined.mem_mode_interleave.vcpu_auto: PASS (39.46 s)
 (04/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_defined.mem_mode_interleave.vcpu_static: PASS (41.77 s)
 (05/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_defined.mem_mode_preferred.vcpu_auto: PASS (39.09 s)
 (06/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_defined.mem_mode_preferred.vcpu_static: PASS (41.62 s)
 (07/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_defined.mem_mode_restrictive.vcpu_auto: PASS (40.09 s)
 (08/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_defined.mem_mode_restrictive.vcpu_static: PASS (41.99 s)
 (09/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_undefined.mem_mode_none.vcpu_auto: PASS (38.60 s)
 (10/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_undefined.mem_mode_none.vcpu_static: PASS (41.63 s)
 (11/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_undefined.mem_mode_strict.vcpu_auto: PASS (39.59 s)
 (12/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_undefined.mem_mode_strict.vcpu_static: PASS (6.06 s)
 (13/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_undefined.mem_mode_interleave.vcpu_auto: PASS (39.52 s)
 (14/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_undefined.mem_mode_interleave.vcpu_static: PASS (6.00 s)
 (15/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_undefined.mem_mode_preferred.vcpu_auto: PASS (39.42 s)
 (16/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_undefined.mem_mode_preferred.vcpu_static: PASS (6.00 s)
 (17/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_undefined.mem_mode_restrictive.vcpu_auto: PASS (39.81 s)
 (18/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_undefined.mem_mode_restrictive.vcpu_static: PASS (5.81 s)
 (19/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_defined.mem_mode_strict.vcpu_auto: PASS (39.72 s)
 (20/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_defined.mem_mode_strict.vcpu_static: PASS (44.02 s)
 (21/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_defined.mem_mode_interleave.vcpu_auto: PASS (38.74 s)
 (22/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_defined.mem_mode_interleave.vcpu_static: PASS (42.14 s)
 (23/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_defined.mem_mode_preferred.vcpu_auto: PASS (41.02 s)
 (24/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_defined.mem_mode_preferred.vcpu_static: PASS (41.85 s)
 (25/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_defined.mem_mode_restrictive.vcpu_auto: PASS (39.78 s)
 (26/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_defined.mem_mode_restrictive.vcpu_static: PASS (42.23 s)
 (27/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_undefined.mem_mode_none.vcpu_auto: PASS (38.80 s)
 (28/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_undefined.mem_mode_none.vcpu_static: PASS (41.82 s)
 (29/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_undefined.mem_mode_strict.vcpu_auto: PASS (38.99 s)
 (30/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_undefined.mem_mode_strict.vcpu_static: PASS (6.00 s)
 (31/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_undefined.mem_mode_interleave.vcpu_auto: PASS (39.69 s)
 (32/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_undefined.mem_mode_interleave.vcpu_static: PASS (6.01 s)
 (33/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_undefined.mem_mode_preferred.vcpu_auto: PASS (39.73 s)
 (34/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_undefined.mem_mode_preferred.vcpu_static: PASS (5.74 s)
 (35/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_undefined.mem_mode_restrictive.vcpu_auto: PASS (39.72 s)
 (36/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_undefined.mem_mode_restrictive.vcpu_static: PASS (6.00 s)
RESULTS    : PASS 36 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/lib/avocado/job-results/job-2023-07-19T02.26-b33ecfb/results.html
JOB TIME   : 1180.94 s

`

**Comment by dzhengfy** [comment]
libvirt 8.0.0

```
 (01/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_defined.mem_mode_strict.vcpu_auto: WARN: Test passed but there were warnings during execution. Check the log for details. (33.74 s)
 (02/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_defined.mem_mode_strict.vcpu_static: WARN: Test passed but there were warnings during execution. Check the log for details. (30.11 s)
 (03/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_defined.mem_mode_interleave.vcpu_auto: **FAIL: Expect cpuset.mems=, but found 0-1** (36.46 s)
 (04/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_defined.mem_mode_interleave.vcpu_static: **FAIL: Expect cpuset.mems=, but found 0-1** (35.13 s)
 (05/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_defined.mem_mode_preferred.vcpu_auto: **FAIL: Expect cpuset.mems=, but found 0-1** (36.17 s)
 (06/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_defined.mem_mode_preferred.vcpu_static: **FAIL: Expect cpuset.mems=, but found 0-1** (34.13 s)
 (07/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_defined.mem_mode_restrictive.vcpu_auto: **FAIL: Expect nodeset to be [0], but found [0, 1]** (35.97 s)
 (08/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_defined.mem_mode_restrictive.vcpu_static: **FAIL: Expect nodeset to be [0], but found [0, 1]** (34.74 s)
 (09/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_undefined.mem_mode_none.vcpu_auto: WARN: Test passed but there were warnings during execution. Check the log for details. (32.19 s)
 (10/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_undefined.mem_mode_none.vcpu_static: **FAIL: Expect cpuset.mems=, but found 0-1** (34.49 s)
 (11/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_undefined.mem_mode_strict.vcpu_auto: WARN: Test passed but there were warnings during execution. Check the log for details. (32.22 s)
 (12/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_undefined.mem_mode_strict.vcpu_static: WARN: Test passed but there were warnings during execution. Check the log for details. (5.91 s)
 (13/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_undefined.mem_mode_interleave.vcpu_auto: **FAIL: Expect cpuset.mems=, but found 0-1** (37.59 s)
 (14/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_undefined.mem_mode_interleave.vcpu_static: WARN: Test passed but there were warnings during execution. Check the log for details. (5.64 s)
 (15/36)type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_undefined.mem_mode_preferred.vcpu_auto: CANCEL: command: '/usr/bin/virsh start avocado-vt-vm1 '\nexit_status: 1\nduration: 2.3091739930096082\ninterrupted: False\npid: 72372\nencoding: 'UTF-8'\nstdout: '\n'\nstderr: "error: Failed to start domain 'avocado-vt-vm1'\nerror: internal error: Process exited prior t... (8.36 s)
 (16/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_undefined.mem_mode_preferred.vcpu_static: WARN: Test passed but there were warnings during execution. Check the log for details. (5.69 s)
 (17/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_undefined.mem_mode_restrictive.vcpu_auto: WARN: Test passed but there were warnings during execution. Check the log for details. (32.08 s)
 (18/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.without_iothread.nodeset_undefined.mem_mode_restrictive.vcpu_static: WARN: Test passed but there were warnings during execution. Check the log for details. (6.14 s)
 (19/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_defined.mem_mode_strict.vcpu_auto: WARN: Test passed but there were warnings during execution. Check the log for details. (32.51 s)
 (20/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_defined.mem_mode_strict.vcpu_static: WARN: Test passed but there were warnings during execution. Check the log for details. (31.99 s)
 (21/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_defined.mem_mode_interleave.vcpu_auto: **FAIL: Expect cpuset.mems=, but found 0-1** (37.64 s)
 (22/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_defined.mem_mode_interleave.vcpu_static: **FAIL: Expect cpuset.mems=, but found 0-1** (34.31 s)
 (23/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_defined.mem_mode_preferred.vcpu_auto: **FAIL: Expect cpuset.mems=, but found 0-1** (37.25 s)
 (24/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_defined.mem_mode_preferred.vcpu_static: **FAIL: Expect cpuset.mems=, but found 0-1** (35.07 s)
 (25/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_defined.mem_mode_restrictive.vcpu_auto: F**AIL: Expect nodeset to be [0], but found [0, 1]** (35.97 s)
 (26/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_defined.mem_mode_restrictive.vcpu_static: **FAIL: Expect nodeset to be [0], but found [0, 1]** (36.03 s)
 (27/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_undefined.mem_mode_none.vcpu_auto: WARN: Test passed but there were warnings during execution. Check the log for details. (33.35 s)
 (28/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_undefined.mem_mode_none.vcpu_static: FAIL: Expect cpuset.mems=, but found 0-1 (35.32 s)
 (29/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_undefined.mem_mode_strict.vcpu_auto: WARN: Test passed but there were warnings during execution. Check the log for details. (32.34 s)
 (30/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_undefined.mem_mode_strict.vcpu_static: WARN: Test passed but there were warnings during execution. Check the log for details. (5.91 s)
 (31/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_undefined.mem_mode_interleave.vcpu_auto: **FAIL: Expect cpuset.mems=, but found 0-1** (36.35 s)
 (32/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_undefined.mem_mode_interleave.vcpu_static: WARN: Test passed but there were warnings during execution. Check the log for details. (5.93 s)
 (33/36)CANCEL: command: '/bin/virsh start avocado-vt-vm1 '\nexit_status: 1\nduration: 2.3162900670140516\ninterrupted: False\npid: 73557\nencoding: 'UTF-8'\nstdout: '\n'\nstderr: "error: Failed to start domain 'avocado-vt-vm1'\nerror: internal error: Process exited prior to ex... (8.46 s)
 (34/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_undefined.mem_mode_preferred.vcpu_static: WARN: Test passed but there were warnings during execution. Check the log for details. (5.94 s)
 (35/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_undefined.mem_mode_restrictive.vcpu_auto: WARN: Test passed but there were warnings during execution. Check the log for details. (32.27 s)
 (36/36) type_specific.io-github-autotest-libvirt.guest_numa_node_tuning.auto_memory_nodeset_placement.with_iothread.nodeset_undefined.mem_mode_restrictive.vcpu_static: WARN: Test passed but there were warnings during execution. Check the log for details. (6.17 s)
RESULTS    : PASS 0 | ERROR 0 | FAIL 16 | SKIP 0 | WARN 18 | INTERRUPT 0 | CANCEL 2
JOB HTML   : /var/lib/avocado/job-results/job-2023-07-19T02.47-0bb0430/results.html
JOB TIME   : 957.75 s
```

**Comment by dzhengfy** [comment]
@nanli1 could you help review it? thanks

File: libvirt/tests/src/numa/guest_numa_node_tuning/auto_memory_nodeset_placement.py [hunk 1]
```diff
@@ -0,0 +1,628 @@
...
+    """
+    mem_mode = test_obj.params.get('mem_mode')
+    nodeset = test_obj.params.get('nodeset')
+    vcpu_placement = eval(test_obj.params.get('vm_attrs')).get('placement')
+    test_obj.virsh_dargs.update({'ignore_status': False})
+    vmxml = vm_xml.VMXML.new_from_dumpxml(test_obj.vm.name)
+
+    if vcpu_placement == 'static' and mem_mode == 'none':
+        xpaths_text = [{'element_attrs': ["./numatune"]}]
+        if libvirt_vmxml.check_guest_xml_by_xpaths(vmxml, xpaths_text, ignore_status=True):
```
Comment (nanli1) [inline libvirt/tests/src/numa/guest_numa_node_tuning/auto_memory_nodeset_placement.py:314]: `libvirt_vmxml.check_guest_xml_by_xpaths` has only two params , Does the ` ignore_status=True` work?

**Review by Yingshun — APPROVED** [review]
LGTM

File: libvirt/tests/src/numa/guest_numa_node_tuning/auto_memory_nodeset_placement.py [hunk 2]
```diff
@@ -0,0 +1,628 @@
...
+#
+#   SPDX-License-Identifier: GPL-2.0
+#
+#   Author: Dan Zheng <dzheng@redhat.com>
+#
+
+
+import json
+import os
+import re
```
Comment (nanli1) [inline libvirt/tests/src/numa/guest_numa_node_tuning/auto_memory_nodeset_placement.py:14]: json should be the last in this order

File: libvirt/tests/src/numa/guest_numa_node_tuning/auto_memory_nodeset_placement.py (hunk 1 above)
Comment (dzhengfy) [reply libvirt/tests/src/numa/guest_numa_node_tuning/auto_memory_nodeset_placement.py:314]: https://github.com/avocado-framework/avocado-vt/pull/3716 added new param

File: libvirt/tests/src/numa/guest_numa_node_tuning/auto_memory_nodeset_placement.py (hunk 2 above)
Comment (dzhengfy) [reply libvirt/tests/src/numa/guest_numa_node_tuning/auto_memory_nodeset_placement.py:14]: What is the reason?

**Review by nanli1 — APPROVED** [review]

**Pull request merged**